tests/conftest.py

```diff
@@ -6,3 +6,29 @@
 def fetch_state_result():
     with temporary_settings(updates={PREFECT_ASYNC_FETCH_STATE_RESULT: True}):
         yield
+
+
+# this crashes the tests "test_failing_flow_run";
+# all tests pass without this fixture so will comment
+# this out and defer for later investigation since
+# users will not be using this in their code
+# more info on the error below:
+# https://github.com/PrefectHQ/prefect-dask/issues/55
+# @pytest.fixture(scope="session", autouse=True)
```

Comment:
does changing the `scope` effect the failure at all? what happens if we use the fixture per `function` instead?

Reply:
Same error.

Reply:
If I remove autouse, it works locally, but that's the same as commenting it out.

![image](https://user-images.githubusercontent.com/15331990/205351664-7a8a6132-3afd-4b0e-bf90-22ec1bfa6ce9.png)

Reply:
I wonder if this is related to https://github.com/PrefectHQ/prefect/issues/7720#issuecomment-1335572194 (threads gone awry?) since when I remove run_sync_in_worker_thread, I get:
```
sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) no such table: flow
[SQL: INSERT INTO flow (id, created, updated, name, tags) VALUES (:id, :created, :updated, :name, :tags) ON CONFLICT (name) DO NOTHING]
[parameters: {'id': 'de2902b6-f05b-4a64-8b0b-2886c09cbf47', 'created': '2022-12-02 01:13:50.268951', 'updated': '2022-12-02 01:13:50.268957', 'name': 'test-flow', 'tags': '[]'}]
(Background on this error at: https://sqlalche.me/e/14/e3q8)
```

Reply:
Oo it could be. We've got some sort of really weird issue when there is an event loop managed by someone else in the mix.

Reply:
Another fixture above seems to also manipulate the Prefect settings in the same way as `prefect_test_harness`.

```
def fetch_state_result():
    with temporary_settings(updates={PREFECT_ASYNC_FETCH_STATE_RESULT: True}):
        yield
```

Is it possible we're seeing a conflict issue between the two fixtures?

Reply:
I tried commenting that one out too, but tests fail locally.

Reply:
When I saved `orion.db` to a local directory (rather than temporary), I noticed that it's empty:
![image](https://user-images.githubusercontent.com/15331990/205370145-8333d09e-1ae7-4ab1-a380-98d5791acc0c.png)

I think when one thread shuts down, it clears the database; then the second thread tries to access it, but it's already gone. Interesting that the `processes` pool works, but not `thread` so I wonder if there's some logic in orion that handles processes separately?

Reply:
I tried doing the same thing in flight:
```
def alembic_upgrade(revision: str = "head", dry_run: bool = False):
    """
    Run alembic upgrades on Orion database

    Args:
        revision: The revision passed to `alembic downgrade`. Defaults to 'head', upgrading all revisions.
        dry_run: Show what migrations would be made without applying them. Will emit sql statements to stdout.
    """
    # lazy import for performance
    import alembic.command
    import pandas as pd
    import sqlite3
    with sqlite3.Connection("/Users/andrew/Applications/python/prefect-dask/orion.db") as conn:
        df = pd.read_sql("SELECT * from sqlite_master", conn)
        print(df)

    alembic.command.upgrade(alembic_config(), revision, sql=dry_run)
```



Results in 3 prints of `df`. I expect 4 since there's four tasks: 
```

tests/test_task_runners.py::TestDaskTaskRunner::test_failing_flow_run[dask_task_runner_with_thread_pool] <- ../../../mambaforge/envs/prefect-dask/lib/python3.9/site-packages/prefect/testing/standard_test_suites/task_runners.py sqlite+aiosqlite:////Users/andrew/Applications/python/prefect-dask/orion.db HELLO



      type                                               name         tbl_name  rootpage                                                sql
0    table                                    alembic_version  alembic_version         2  CREATE TABLE alembic_version (\n\tversion_num ...
1    index                 sqlite_autoindex_alembic_version_1  alembic_version         3                                               None
2    table                                               flow             flow         4  CREATE TABLE flow (\n\tid CHAR(36) DEFAULT (\n...
3    index                            sqlite_autoindex_flow_1             flow         5                                               None
4    index                            sqlite_autoindex_flow_2             flow         6                                               None
..     ...                                                ...              ...       ...                                                ...
122  index                       ix_flow_run__state_timestamp         flow_run        82  CREATE INDEX ix_flow_run__state_timestamp ON f...
123  index                       ix_task_run__state_timestamp         task_run        97  CREATE INDEX ix_task_run__state_timestamp ON t...
124  index                             ix_deployment__created       deployment        83  CREATE INDEX ix_deployment__created ON deploym...
125  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       137  CREATE INDEX ix_flow_run__coalesce_start_time_...
126  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       138  CREATE INDEX ix_flow_run__coalesce_start_time_...

[127 rows x 5 columns]
11:28:31.742 | INFO    | prefect.engine - Created flow run 'colossal-swine' for flow 'test-flow'
11:28:31.743 | INFO    | prefect.task_runner.dask - Creating a new Dask cluster with `distributed.deploy.local.LocalCluster`
11:28:32.028 | INFO    | prefect.task_runner.dask - The Dask dashboard is available at http://192.168.50.68:8787/status
11:28:32.129 | INFO    | Flow run 'colossal-swine' - Created task run 'task_a-4179b103-0' for task 'task_a'
11:28:32.435 | INFO    | Flow run 'colossal-swine' - Submitted task run 'task_a-4179b103-0' for execution.
      type                                               name         tbl_name  rootpage                                                sql
0    table                                    alembic_version  alembic_version         2  CREATE TABLE alembic_version (\n\tversion_num ...
1    index                 sqlite_autoindex_alembic_version_1  alembic_version         3                                               None
2    table                                               flow             flow         4  CREATE TABLE flow (\n\tid CHAR(36) DEFAULT (\n...
3    index                            sqlite_autoindex_flow_1             flow         5                                               None
4    index                            sqlite_autoindex_flow_2             flow         6                                               None
..     ...                                                ...              ...       ...                                                ...
122  index                       ix_flow_run__state_timestamp         flow_run        82  CREATE INDEX ix_flow_run__state_timestamp ON f...
123  index                       ix_task_run__state_timestamp         task_run        97  CREATE INDEX ix_task_run__state_timestamp ON t...
124  index                             ix_deployment__created       deployment        83  CREATE INDEX ix_deployment__created ON deploym...
125  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       137  CREATE INDEX ix_flow_run__coalesce_start_time_...
126  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       138  CREATE INDEX ix_flow_run__coalesce_start_time_...

[127 rows x 5 columns]
11:28:32.458 | INFO    | Flow run 'colossal-swine' - Created task run 'task_b-8e3be95f-0' for task 'task_b'
11:28:32.460 | INFO    | Flow run 'colossal-swine' - Submitted task run 'task_b-8e3be95f-0' for execution.
      type                                               name         tbl_name  rootpage                                                sql
0    table                                    alembic_version  alembic_version         2  CREATE TABLE alembic_version (\n\tversion_num ...
1    index                 sqlite_autoindex_alembic_version_1  alembic_version         3                                               None
2    table                                               flow             flow         4  CREATE TABLE flow (\n\tid CHAR(36) DEFAULT (\n...
3    index                            sqlite_autoindex_flow_1             flow         5                                               None
4    index                            sqlite_autoindex_flow_2             flow         6                                               None
..     ...                                                ...              ...       ...                                                ...
122  index                       ix_flow_run__state_timestamp         flow_run        82  CREATE INDEX ix_flow_run__state_timestamp ON f...
123  index                       ix_task_run__state_timestamp         task_run        97  CREATE INDEX ix_task_run__state_timestamp ON t...
124  index                             ix_deployment__created       deployment        83  CREATE INDEX ix_deployment__created ON deploym...
125  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       137  CREATE INDEX ix_flow_run__coalesce_start_time_...
126  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       138  CREATE INDEX ix_flow_run__coalesce_start_time_...

[127 rows x 5 columns]
11:28:32.497 | INFO    | Flow run 'colossal-swine' - Created task run 'task_c-167cbbb3-0' for task 'task_c'
11:28:32.499 | INFO    | Flow run 'colossal-swine' - Submitted task run 'task_c-167cbbb3-0' for execution.
11:28:32.521 | INFO    | Flow run 'colossal-swine' - Created task run 'task_c-167cbbb3-1' for task 'task_c'
11:28:32.523 | INFO    | Flow run 'colossal-swine' - Submitted task run 'task_c-167cbbb3-1' for execution.
2022-12-02 11:28:32,546 - distributed.worker - WARNING - Compute Failed
Key:       task_b-8e3be95f-0-ede277d99dc0460da38d7f6c6446d2af-1
Function:  run
args:      ()
kwargs:    {}
Exception: "KeyError('config')"

```
Reaching the 4th (task_c-167cbbb3-1 vs task_c-167cbbb3-**0**), it doesn't print because I don't think the database exists anymore?

Reply:
If I wrap a try/except KeyError, it passes and all the dataframes show as expected:
```
def alembic_upgrade(revision: str = "head", dry_run: bool = False):

    # lazy import for performance
    import alembic.command
    import pandas as pd
    import sqlite3
    with sqlite3.Connection("/Users/andrew/Applications/python/prefect-dask/orion.db") as conn:
        df = pd.read_sql("SELECT * from sqlite_master", conn)
        print(df)

    try:
        alembic.command.upgrade(alembic_config(), revision, sql=dry_run)
    except KeyError:
        pass

```

```

      type                                               name         tbl_name  rootpage                                                sql
0    table                                    alembic_version  alembic_version         2  CREATE TABLE alembic_version (\n\tversion_num ...
1    index                 sqlite_autoindex_alembic_version_1  alembic_version         3                                               None
2    table                                               flow             flow         4  CREATE TABLE flow (\n\tid CHAR(36) DEFAULT (\n...
3    index                            sqlite_autoindex_flow_1             flow         5                                               None
4    index                            sqlite_autoindex_flow_2             flow         6                                               None
..     ...                                                ...              ...       ...                                                ...
122  index                       ix_flow_run__state_timestamp         flow_run        82  CREATE INDEX ix_flow_run__state_timestamp ON f...
123  index                       ix_task_run__state_timestamp         task_run        97  CREATE INDEX ix_task_run__state_timestamp ON t...
124  index                             ix_deployment__created       deployment        83  CREATE INDEX ix_deployment__created ON deploym...
125  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       137  CREATE INDEX ix_flow_run__coalesce_start_time_...
126  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       138  CREATE INDEX ix_flow_run__coalesce_start_time_...

[127 rows x 5 columns]
11:34:30.635 | INFO    | prefect.engine - Created flow run 'opalescent-ammonite' for flow 'test-flow'
11:34:30.636 | INFO    | prefect.task_runner.dask - Creating a new Dask cluster with `distributed.deploy.local.LocalCluster`
11:34:30.851 | INFO    | prefect.task_runner.dask - The Dask dashboard is available at http://192.168.50.68:8787/status
11:34:30.923 | INFO    | Flow run 'opalescent-ammonite' - Created task run 'task_a-4179b103-0' for task 'task_a'
11:34:31.187 | INFO    | Flow run 'opalescent-ammonite' - Submitted task run 'task_a-4179b103-0' for execution.
      type                                               name         tbl_name  rootpage                                                sql
0    table                                    alembic_version  alembic_version         2  CREATE TABLE alembic_version (\n\tversion_num ...
1    index                 sqlite_autoindex_alembic_version_1  alembic_version         3                                               None
2    table                                               flow             flow         4  CREATE TABLE flow (\n\tid CHAR(36) DEFAULT (\n...
3    index                            sqlite_autoindex_flow_1             flow         5                                               None
4    index                            sqlite_autoindex_flow_2             flow         6                                               None
..     ...                                                ...              ...       ...                                                ...
122  index                       ix_flow_run__state_timestamp         flow_run        82  CREATE INDEX ix_flow_run__state_timestamp ON f...
123  index                       ix_task_run__state_timestamp         task_run        97  CREATE INDEX ix_task_run__state_timestamp ON t...
124  index                             ix_deployment__created       deployment        83  CREATE INDEX ix_deployment__created ON deploym...
125  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       137  CREATE INDEX ix_flow_run__coalesce_start_time_...
126  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       138  CREATE INDEX ix_flow_run__coalesce_start_time_...

[127 rows x 5 columns]
11:34:31.204 | INFO    | Flow run 'opalescent-ammonite' - Created task run 'task_b-8e3be95f-0' for task 'task_b'
11:34:31.207 | INFO    | Flow run 'opalescent-ammonite' - Submitted task run 'task_b-8e3be95f-0' for execution.
      type                                               name         tbl_name  rootpage                                                sql
0    table                                    alembic_version  alembic_version         2  CREATE TABLE alembic_version (\n\tversion_num ...
1    index                 sqlite_autoindex_alembic_version_1  alembic_version         3                                               None
2    table                                               flow             flow         4  CREATE TABLE flow (\n\tid CHAR(36) DEFAULT (\n...
3    index                            sqlite_autoindex_flow_1             flow         5                                               None
4    index                            sqlite_autoindex_flow_2             flow         6                                               None
..     ...                                                ...              ...       ...                                                ...
122  index                       ix_flow_run__state_timestamp         flow_run        82  CREATE INDEX ix_flow_run__state_timestamp ON f...
123  index                       ix_task_run__state_timestamp         task_run        97  CREATE INDEX ix_task_run__state_timestamp ON t...
124  index                             ix_deployment__created       deployment        83  CREATE INDEX ix_deployment__created ON deploym...
125  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       137  CREATE INDEX ix_flow_run__coalesce_start_time_...
126  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       138  CREATE INDEX ix_flow_run__coalesce_start_time_...

[127 rows x 5 columns]
11:34:31.238 | INFO    | Flow run 'opalescent-ammonite' - Created task run 'task_c-167cbbb3-0' for task 'task_c'
11:34:31.240 | INFO    | Flow run 'opalescent-ammonite' - Submitted task run 'task_c-167cbbb3-0' for execution.
11:34:31.259 | INFO    | Flow run 'opalescent-ammonite' - Created task run 'task_c-167cbbb3-1' for task 'task_c'
11:34:31.261 | INFO    | Flow run 'opalescent-ammonite' - Submitted task run 'task_c-167cbbb3-1' for execution.
11:34:31.291 | ERROR   | Task run 'task_a-4179b103-0' - Encountered exception during execution:
Traceback (most recent call last):
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/prefect/engine.py", line 1346, in orchestrate_task_run
    result = await run_sync(task.fn, *args, **kwargs)
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/prefect/utilities/asyncutils.py", line 69, in run_sync_in_worker_thread
    return await anyio.to_thread.run_sync(call, cancellable=True)
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/prefect/testing/standard_test_suites/task_runners.py", line 76, in task_a
    raise RuntimeError("This task fails!")
RuntimeError: This task fails!
11:34:31.300 | ERROR   | Task run 'task_b-8e3be95f-0' - Encountered exception during execution:
Traceback (most recent call last):
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/prefect/engine.py", line 1346, in orchestrate_task_run
    result = await run_sync(task.fn, *args, **kwargs)
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/prefect/utilities/asyncutils.py", line 69, in run_sync_in_worker_thread
    return await anyio.to_thread.run_sync(call, cancellable=True)
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/anyio/to_thread.py", line 31, in run_sync
    return await get_asynclib().run_sync_in_worker_thread(
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 937, in run_sync_in_worker_thread
    return await future
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/anyio/_backends/_asyncio.py", line 867, in run
    result = context.run(func, *args)
  File "/Users/andrew/mambaforge/envs/prefect-dask/lib/python3.9/site-packages/prefect/testing/standard_test_suites/task_runners.py", line 80, in task_b
    raise ValueError("This task fails and passes data downstream!")
ValueError: This task fails and passes data downstream!
11:34:31.308 | ERROR   | Task run 'task_a-4179b103-0' - Finished in state Failed('Task run encountered an exception: RuntimeError: This task fails!\n')
11:34:31.315 | ERROR   | Task run 'task_b-8e3be95f-0' - Finished in state Failed('Task run encountered an exception: ValueError: This task fails and passes data downstream!\n')
      type                                               name         tbl_name  rootpage                                                sql
0    table                                    alembic_version  alembic_version         2  CREATE TABLE alembic_version (\n\tversion_num ...
1    index                 sqlite_autoindex_alembic_version_1  alembic_version         3                                               None
2    table                                               flow             flow         4  CREATE TABLE flow (\n\tid CHAR(36) DEFAULT (\n...
3    index                            sqlite_autoindex_flow_1             flow         5                                               None
4    index                            sqlite_autoindex_flow_2             flow         6                                               None
..     ...                                                ...              ...       ...                                                ...
122  index                       ix_flow_run__state_timestamp         flow_run        82  CREATE INDEX ix_flow_run__state_timestamp ON f...
123  index                       ix_task_run__state_timestamp         task_run        97  CREATE INDEX ix_task_run__state_timestamp ON t...
124  index                             ix_deployment__created       deployment        83  CREATE INDEX ix_deployment__created ON deploym...
125  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       137  CREATE INDEX ix_flow_run__coalesce_start_time_...
126  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       138  CREATE INDEX ix_flow_run__coalesce_start_time_...

[127 rows x 5 columns]
      type                                               name         tbl_name  rootpage                                                sql
0    table                                    alembic_version  alembic_version         2  CREATE TABLE alembic_version (\n\tversion_num ...
1    index                 sqlite_autoindex_alembic_version_1  alembic_version         3                                               None
2    table                                               flow             flow         4  CREATE TABLE flow (\n\tid CHAR(36) DEFAULT (\n...
3    index                            sqlite_autoindex_flow_1             flow         5                                               None
4    index                            sqlite_autoindex_flow_2             flow         6                                               None
..     ...                                                ...              ...       ...                                                ...
122  index                       ix_flow_run__state_timestamp         flow_run        82  CREATE INDEX ix_flow_run__state_timestamp ON f...
123  index                       ix_task_run__state_timestamp         task_run        97  CREATE INDEX ix_task_run__state_timestamp ON t...
124  index                             ix_deployment__created       deployment        83  CREATE INDEX ix_deployment__created ON deploym...
125  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       137  CREATE INDEX ix_flow_run__coalesce_start_time_...
126  index  ix_flow_run__coalesce_start_time_expected_star...         flow_run       138  CREATE INDEX ix_flow_run__coalesce_start_time_...

[127 rows x 5 columns]
11:34:31.476 | ERROR   | Flow run 'opalescent-ammonite' - Finished in state Failed('2/4 states failed.')
PASSED
```

Reply:
If I drop "run_sync_in_worker_thread", this works
```
    async def run_migrations_upgrade(self):
        """Run all upgrade migrations"""
        alembic_upgrade()
```